### ISSUE TITLE
Isolate global styles

### ISSUE THREAD

**Issue by xaddict** [histoire-dev/histoire#339]
### Describe the bug

When using a histoire.setup.js or style block inside a Vue 3 SFC to import global styles, these styles bleed out into the Histoire styling, overriding color, inputs and theming.

I don't think my application styling should inform Histoire's styling, right?

I suggest to:
- support the [scoped] attribute on style blocks inside SFCs
- to make the whole Histoire render area an iframe no matter the settings for grid/single
- to support a key `styles` in histoire.config.js

### Reproduction

- Create an Histoire Vue 3 project
- Create a style sheet with `*{color: black}`
- Import the style sheet in histoire.setup.js
- Turn on dark mode
- See how the style sheet overrides Histoire's dark theme.

https://stackblitz.com/edit/histoire-vue3-starter-8dm7mb?file=src/index.css

### System Info

```shell
System:
    OS: macOS 12.6
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 56.88 MB / 16.00 GB
    Shell: 5.8.1 - /bin/zsh
  Binaries:
    Node: 18.7.0 - ~/.nvm/versions/node/v18.7.0/bin/node
    Yarn: 1.22.19 - ~/.yarn/bin/yarn
    npm: 8.15.0 - ~/.nvm/versions/node/v18.7.0/bin/npm
  Browsers:
    Brave Browser: 102.1.39.111
    Chrome: 106.0.5249.119
    Edge: 106.0.1370.52
    Firefox: 106.0.1
    Firefox Developer Edition: 107.0
    Firefox Nightly: 107.0a1
    Safari: 16.0
    Safari Technology Preview: 16.4
  npmPackages:
    @histoire/plugin-vue: ^0.11.5 => 0.11.5 
    @vitejs/plugin-vue: ^3.0.3 => 3.1.2 
    histoire: ^0.11.5 => 0.11.5 
    vite: ^3.0.9 => 3.1.8
```

### Used Package Manager

yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/histoire-dev/histoire/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/histoire-dev/histoire/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://histoire.dev/guide/).
- [X] Check that there isn't [already an issue](https://github.com/histoire-dev/histoire/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/histoire-dev/histoire/discussions).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by tillsanders** [comment]
I have the same issue: https://github.com/histoire-dev/histoire/issues/66#issuecomment-1282343948

I would love to see this be addressed in some way. I'm open to sponsoring a little to help make this happen :)

**Comment by troyere** [comment]
Not strictly related, but there's no iframe **at all** in grid mode.

**Comment by Akryum** [comment]
> support the [scoped] attribute on style blocks inside SFCs

This is already supported :)

**Comment by xaddict** [comment]
Yes that’s not the problem. Global styles leaking into histoire’s UI are the problem

Met vriendelijke groet,
Luuk Lamers

**Comment by Akryum** [comment]
Yeah it's a tough issue with no obvious solution (iframe comes with a lot of tradeoffs)

**Comment by tillsanders** [comment]
Would it be possible to run the stories inside a shadow DOM? That would isolate the styles completely.

**Comment by xaddict** [comment]
I don't think so, as I have a global stylesheet that gets imported through the main javascript file. I can't import the stylesheet in every story.

**Comment by PhotonBursted** [comment]
After playing around a bit I noticed some things that may help this along. Perhaps sharing them sparks some new inspiration to get this resolved.

The styles imported in `histoire.setup.js` are included in the main Histoire app, as well as in the sandbox. It looks like they are "lazily loaded"; the styles from the sandbox only start "bleeding" once the first story is opened. The sandbox itself is an iframe which contains a rendering of the app with the styles included separately.

This can be confirmed by opening the Stackblitz in the original post, opening the app in a new tab, and playing around with it there. It starts out with three `<style>` tags in the app's `<head>`, which turns into four once a story is opened. After manually deleting the last one the Histoire app is perfectly displayed, while the styles in the sandbox are retained.

I guess this means the feature is somehow halfway there already? It definitely seems to be possible, at least.

**Comment by Aksamyt** [comment]
This is an issue that prevents histoire from being used in a ton of projects, unfortunately. I hope this’ll be solved soon.

Thanks to @PhotonBursted’s comment, I wrote this in my `histoire.setup.ts` file and it seems stable enough:
```ts
const isIframe = window.self !== window.top;
document.head
    .querySelectorAll("style[type='text/css']:not([data-vite-dev-id*='histoire'])")
    .forEach((style) => isIframe || style.remove());
```
It’s dirty but it works with v0.15.9

**Comment by reslear** [comment]
@Aksamyt but work only in dev mode

**Comment by t0byman** [comment]
Would love a fix that works for production

**Comment by noclat** [comment]
Hi, any updates about isolating Histoire UI and global story styles? That issue defeats Histoire's purpose (sandboxing).

**Comment by skstuder** [comment]
Was so excited at how fast we got this working, but now I am embarrassed to show anyone how bad it looks, because the global styles leak and make Histoire look terrible.

**Comment by SteinRobert** [comment]
> ```ts
> const isIframe = window.self !== window.top;
> document.head
>     .querySelectorAll("style[type='text/css']:not([data-vite-dev-id*='histoire'])")
>     .forEach((style) => isIframe || style.remove());
> ```

Also hacky - however works for now - took this and added it to setupCode:
```js
export default defineConfig({
  plugins: [
    HstVue(),
    HstNuxt(),
  ],
  setupCode: [`const isIframe = window.self !== window.top;
  document.head
      .querySelectorAll("style[type='text/css']:not([data-vite-dev-id*='histoire'])")
      .forEach((style) => isIframe || style.remove());`]
})
```

**Comment by kuttyhub** [comment]
Hi, any updates about this issue?

**Comment by romeromembersy** [comment]
After almost 3 years this is still an issue :(
I know this is an open source project, and I'm thankful for @Akryum work, but this issue makes this project literally unusable.
Histoire's own styles are overriding the style of the components I'm trying to showcase, it defeats the purpose of the tool itself.

**Comment by reslear** [comment]
@romeromembersy i think he lost interest in this a long time ago, alternatives no good solutions but as usual 
just migrate to storybook (no vue sfc) or just vitepress or solutions with nuxt ui (no dynamic props)

**Comment by romeromembersy** [comment]
@reslear I get it, I have my own unmaintained projects...
I moved to Vitepress

**Comment by mtancoigne** [comment]
I ended by writing [my own "storybook-like"](https://rise-ui.experimentslabs.com/) documentation system for a components library (for fun and to try something new).

The downside is that you need to have a component for every part of the documentation.
The good side is that you have full control on what is done, and can [reuse the documentation system](https://doc.garden-party.io/) in projects using the lib (i.e: a documentation system for Bootstrap, usable to document bootstrap-based projects)

It was a super interesting journey; I first thought it would be super complicated to get something working, but it's not that hard: the proof of concept (md pages, rendering components with a "playground") took ~4h.

It seems OK when you provide a library with a lot of components, but feels like a waste of time for small projects.

Anyway, if you need inspiration, [the code is on Gitlab](https://gitlab.com/experimentslabs/garden-party/rise-ui/-/tree/develop/documentation_system)

**Comment by hugoattal** [comment]
It should be fairly easy to isolate the style now with the [`@scope`](https://developer.mozilla.org/en-US/docs/Web/CSS/@scope) CSS selector

The idea would be to encapsulate all the CSS inside an `@scope`, and replace all the `:root` by [`:scope`](https://developer.mozilla.org/en-US/docs/Web/CSS/:scope)

Thing is, [it's not yet supported by Firefox.](https://caniuse.com/?search=%40scope)..
